README.md
# AnyEvent::Open3::Simple ![linux](https://github.com/uperl/AnyEvent-Open3-Simple/workflows/linux/badge.svg) ![windows](https://github.com/uperl/AnyEvent-Open3-Simple/workflows/windows/badge.svg) ![macos](https://github.com/uperl/AnyEvent-Open3-Simple/workflows/macos/badge.svg) ![msys2-mingw](https://github.com/uperl/AnyEvent-Open3-Simple/workflows/msys2-mingw/badge.svg)

Interface to open3 under AnyEvent

# SYNOPSIS

```perl
use 5.010;
use AnyEvent;
use AnyEvent::Open3::Simple;

my $done = AnyEvent->condvar;

my $ipc = AnyEvent::Open3::Simple->new(
  on_start => sub {
    my $proc = shift;       # isa AnyEvent::Open3::Simple::Process
    my $program = shift;    # string
    my @args = @_;          # list of arguments
    say 'child PID: ', $proc->pid;
  },
  on_stdout => sub {
    my $proc = shift;       # isa AnyEvent::Open3::Simple::Process
    my $line = shift;       # string
    say 'out: ', $string;
  },
  on_stderr => sub {
    my $proc = shift;       # isa AnyEvent::Open3::Simple::Process
    my $line = shift;       # string
    say 'err: ', $line;
  },
  on_exit   => sub {
    my $proc = shift;       # isa AnyEvent::Open3::Simple::Process
    my $exit_value = shift; # integer
    my $signal = shift;     # integer
    say 'exit value: ', $exit_value;
    say 'signal:     ', $signal;
    $done->send;
  },
  on_error => sub {
    my $error = shift;      # the exception thrown by IPC::Open3::open3
    my $program = shift;    # string
    my @args = @_;          # list of arguments
    warn "error: $error";
    $done->send;
  },
);

$ipc->run('echo', 'hello there');
$done->recv;
```

# DESCRIPTION

This module provides an interface to open3 while running under AnyEvent
that delivers data from stdout and stderr as lines are written by the
subprocess.  The interface is reminiscent of [IPC::Open3::Simple](https://metacpan.org/pod/IPC::Open3::Simple),
although this module does provides a somewhat different API, so it
cannot be used a drop in replacement for that module.

There are already a number of interfaces for interacting with subprocesses
in the context of [AnyEvent](https://metacpan.org/pod/AnyEvent), but this one is the most convenient for my
usage.  Note the modules listed in the SEE ALSO section below for other
interfaces that may be more or less appropriate.

# CONSTRUCTOR

Constructor takes a hash or hashref of event callbacks and attributes.
Event callbacks have an `on_` prefix, attributes do not.

## ATTRIBUTES

- implementation

    The implementation to use for detecting process termination.  This should
    be one of `child`, `idle` or `mojo`.  On all platforms except for Microsoft
    Windows (but not Cygwin) the default is `child`.

    You can change the default by setting the `ANYEVENT_OPEN3_SIMPLE`
    environment variable, like this:

    ```
    % export ANYEVENT_OPEN3_SIMPLE=idle
    ```

    The `mojo` implementation is experimental and allows you to use
    [AnyEvent::Open3::Simple](https://metacpan.org/pod/AnyEvent::Open3::Simple) with [Mojolicious](https://metacpan.org/pod/Mojolicious) but without [EV](https://metacpan.org/pod/EV)
    (which is usually required for [AnyEvent](https://metacpan.org/pod/AnyEvent), [Mojolicious](https://metacpan.org/pod/Mojolicious) interaction).

## EVENTS

These events will be triggered by the subprocess when the run method is
called. Each event callback (except `on_error`) gets passed in an
instance of [AnyEvent::Open3::Simple::Process](https://metacpan.org/pod/AnyEvent::Open3::Simple::Process) as its first argument
which can be used to get the PID of the subprocess, or to write to it.
`on_error` does not get a process object because it indicates an error in
the creation of the process.

Not all of these events will fire depending on the execution of the
child process.  In the very least exactly one of `on_start` or `on_error`
will be called.

- `on_start` ($proc, $program, @arguments)

    Called after the process is created, but before the run method returns
    (that is, it does not wait to re-enter the event loop first).

    In versions 0.78 and better, this event also gets the program name
    and arguments passed into the [run](https://metacpan.org/pod/AnyEvent::Open3::Simple#run)
    method.

- `on_error` ($error, $program, @arguments)

    Called when there is an execution error, for example, if you ask
    to run a program that does not exist.  No process is passed in
    because the process failed to create.  The error passed in is
    the error thrown by [IPC::Open3](https://metacpan.org/pod/IPC::Open3) (typically a string which begins
    with "open3: ...").

    In some environments open3 is unable to detect exec errors in the
    child, so you may not be able to rely on this event.  It does
    seem to work consistently on Perl 5.14 or better though.

    Different environments have different ways of handling it when
    you ask to run a program that doesn't exist.  On Linux and Cygwin,
    this will raise an `on_error` event, on `MSWin32` it will
    not trigger a `on_error` and instead cause a normal exit
    with a exit value of 1.

    In versions 0.77 and better, this event also gets the program name
    and arguments passed into the [run](https://metacpan.org/pod/AnyEvent::Open3::Simple#run)
    method.

- `on_stdout` ($proc, $line)

    Called on every line printed to stdout by the child process.

- `on_stderr` ($proc, $line)

    Called on every line printed to stderr by the child process.

- `on_exit` ($proc, $exit\_value, $signal)

    Called when the processes completes, either because it called exit,
    or if it was killed by a signal.

- `on_success` ($proc)

    Called when the process returns zero exit value and is not terminated by a signal.

- `on_signal` ($proc, $signal)

    Called when the processes is terminated by a signal.

- `on_fail` ($proc, $exit\_value)

    Called when the process returns a non-zero exit value.

# METHODS

## run

```perl
$ipc->run($program, @arguments);
$ipc->run($program, @arguments, \$stdin);             # (version 0.76)
$ipc->run($program, @arguments, \@stdin);             # (version 0.76)
$ipc->run($program, @arguments, sub {...});           # (version 0.80)
$ipc->run($program, @arguments, \$stdin, sub {...});  # (version 0.80)
$ipc->run($program, @arguments, \@stdin, sub {...});  # (version 0.80)
```

Start the given program with the given arguments.  Returns
immediately.  Any events that have been specified in the
constructor (except for `on_start`) will not be called until
the process re-enters the event loop.

You may optionally provide the full content of standard input
as a string reference or list reference as the last argument
(or second to last if you are providing a callback below).
If provided as a list reference, it will be joined by new lines
in whatever format is native to your Perl.  Currently on
(non cygwin) Windows (Strawberry, ActiveState) this is the only
way to provide standard input to the subprocess.

Do not mix the use of passing standard input to [run](https://metacpan.org/pod/AnyEvent::Open3::Simple#run)
and [AnyEvent::Open3::Simple::Process#print](https://metacpan.org/pod/AnyEvent::Open3::Simple::Process#print) or [AnyEvent::Open3::Simple::Process#say](https://metacpan.org/pod/AnyEvent::Open3::Simple::Process#say),
otherwise bad things may happen.

In version 0.80 or better, you may provide a callback as the last argument
which is called before `on_start`, and takes the process object as its only
argument.  For example:

```perl
foreach my $i (1..10)
{
  $ipc->run($prog, @args, \$stdin, sub {
    my($proc) = @_;
    $proc->user({ iteration => $i });
  });
}
```

This is useful for making data accessible to `$ipc` object's callbacks that may
be out of scope otherwise.

# CAVEATS

Some AnyEvent implementations may not work properly with the method
used by AnyEvent::Open3::Simple to wait for the child process to
terminate.  See ["CHILD PROCESS WATCHERS" in AnyEvent](https://metacpan.org/pod/AnyEvent#CHILD-PROCESS-WATCHERS) for details.

This module uses an idle watcher instead of a child watcher to detect
program termination on Microsoft Windows (but not Cygwin).  This is
because the child watchers are unsupported by AnyEvent on Windows.
The idle watcher implementation seems to pass the test suite, but there
may be some traps for the unwary.  There may be other platforms or
event loops where this is the appropriate choice, and you can use the
`ANYEVENT_OPEN3_SIMPLE` environment variable or the `implementation`
attribute to force it use an idle watcher instead.  Patches for detecting
environments where idle watchers should be used are welcome and
encouraged.

As of version 0.85, this module works on Windows with [AnyEvent::Impl::EV](https://metacpan.org/pod/AnyEvent::Impl::EV),
[AnyEvent::Impl::Event](https://metacpan.org/pod/AnyEvent::Impl::Event) and [AnyEvent::Impl::Perl](https://metacpan.org/pod/AnyEvent::Impl::Perl) (possibly others),
although in the past they have either not worked or had limitations placed
on them.  Because the author of [AnyEvent](https://metacpan.org/pod/AnyEvent) does not hold the native Windows
port of Perl in high regard: problems such as this may pop up again
in the future and may not be addressed, and may be out of the control of the
author of this module.

Performance for the idle watcher implementation on native Windows (non-Cygwin)
is almost certainly suboptimal, but the author of this module uses it
and finds it useful despite this.

Writing to a subprocesses stdin with [AnyEvent::Open3::Simple::Process#print](https://metacpan.org/pod/AnyEvent::Open3::Simple::Process#print)
or [AnyEvent::Open3::Simple::Process#say](https://metacpan.org/pod/AnyEvent::Open3::Simple::Process#say) is unsupported on Microsoft
Windows (it does work under Cygwin though).

There are some traps for the unwary relating to buffers and deadlocks,
[IPC::Open3](https://metacpan.org/pod/IPC::Open3) is recommended reading.

If you register a call back for `on_exit`, but not `on_error` then
use a condition variable to wait for the process to complete as in
this:

```perl
my $cv = AnyEvent->condvar;
my $ipc = AnyEvent::Open3::Simple->new(
  on_exit => sub { $cv->send },
);
$ipc->run('command_not_found');
$cv->recv;
```

You might be waiting forever if there is an error starting the
process (if for example you give it a bad command).  To handle
this situation you might use croak on the condition variable
in the event of error:

```perl
my $cv = AnyEvent->condvar;
my $ipc = AnyEvent::Open3::Simple->new(
  on_exit => sub { $cv->send },
  on_error => sub {
    my $error = shift;
    $cv->croak($error);
  },
);
$ipc->run('command_not_found');
$cv->recv;
```

This will cause the `recv` to die, printing a useful diagnostic
if the exception isn't caught somewhere else.

# SEE ALSO

- [AnyEvent::Open3::Simple::Process](https://metacpan.org/pod/AnyEvent::Open3::Simple::Process)

    Represents a process being run by this module, typically passed
    into the callbacks.

- [AnyEvent::Subprocess](https://metacpan.org/pod/AnyEvent::Subprocess)

    Alternative to this module.

- [AnyEvent::Run](https://metacpan.org/pod/AnyEvent::Run)

    Alternative to this module.

# AUTHOR

Author: Graham Ollis <plicease@cpan.org>

Contributors:

Stephen R. Scaffidi

Scott Wiersdorf

Graham Knop (HAARG)

# COPYRIGHT AND LICENSE

This software is copyright (c) 2012-2022 by Graham Ollis.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for {{$dist->name}}

{{$NEXT}}

0.90      2023-08-28 07:54:42 -0600
  - Documentation improvements (gh#22)

0.89      2023-08-27 10:47:27 -0600
  - Fix bug that caused test failure in recent bleadperl (gh#20, gh#21)

0.88      2022-10-26 09:01:38 -0600
  - Documentation improvements (gh#19)

0.87      2019-04-16 17:32:23 -0400
  - Require Test::More 0.98 for test fixes (gh#12)

0.86      2016-01-14 09:41:20 -0500
  - Experimental support for Mojolicious without EV (use 'mojo' as implementation)
  - Use IO::Select for actual non-blocking on windows (fixes bug introduced in 0.85)

0.85      2016-01-13 14:59:03 -0500
  - MSWin32 supports EV again.
  - MSWin32 supports AnyEvent::Impl::Perl for the first time.
  - Switched to EUMM for installer

0.84      2015-06-11 10:39:56 -0400
  - Fix accidental breakage in 5.8.8

0.83      2015-05-15 07:16:25 -0400
  - Remove deprecated stdin argument passed into new
  - Remove experimental "raw" support

0.82      2015-03-03 23:37:46 -0500
  - Explicitly require Perl 5.6.0 or better
  - Clean up prereqs

0.81      2014-10-08 11:11:53 -0400
  - use Event instead of EV in testing for Strawberry Perl 5.20 or better
    fixes test failures on Strawberry 5.20.*

0.80      2014-07-22 06:50:13 -0400
  - promote to production

0.79_02   2014-07-19 11:44:10 -0400
  - identical release new version

0.79_01   2014-07-18 11:05:26 -0400
  - added optional callback to run method to allow per-execute callback to
    be called just before on_start.

0.79      2014-04-19 08:25:02 -0400
  - migrate from MakeMaker to Module::Build
  - documentation tweaks

0.78      2014-03-20 18:15:47 -0400
  - on_start event now gets program and arguments passed to the run method

0.77      2014-03-20 17:55:47 -0400
  - on_error event now gets program and arguments passed to the run method
  - add user attribute on AnyEvent::Open3::Simple::Process

0.76      2014-03-18 12:04:18 -0400
  - deprecate stdin attribute for AnyEvent::Open3::Simple
  - optionally pass whole content stdin to run method instead

0.75      2013-09-17 07:54:01 -0400
  - add stdin attribute for AnyEvent::Open3::Simple

0.74      2013-09-12 06:10:50 -0400
  - skip raw support testing on all platforms

0.73      2013-09-10 14:08:24 -0400
  - perl 5.8 support fixes

0.72      2013-09-09 08:20:35 -0400
  - marking raw support as experiemental
    https://github.com/plicease/AnyEvent-Open3-Simple/issues/6
  - undocumented (and experimental): on_stderr and on_stdout can be list ref

0.71      2013-09-03 15:19:59 -0400
  - fixed raw support

0.70      2013-09-03 12:39:38 -0400
  - promote to production

0.69_02   2013-09-01 14:25:50 -0400
  - add prereq EV on MSWin32

0.69_01   2013-08-31 20:28:03 -0400
  - experimental perl 5.8 support
  - added optional raw attribute to constructor

0.69      2013-08-29 21:58:23 -0400
  - added an alternate idle watcher implementation
  - added MSWin32 support (via idle watcher)
  - change end line processing
    in the past we used chomp() which removed trailing $/
    but this didn't work nicely with MSWin32, instead we
    use a regex that removes \r\n, \n or \r

0.68      2013-08-23 08:38:40 -0400
  - documentation tweaks

0.67      2013-03-21 11:01:47 America/New_York
  - fix META.json,yml

0.66      2013-03-07 08:39:25 America/New_York
  - added on_success callback

0.65      2013-02-09 17:10:56 America/New_York
  - documentation updates

0.64      2013-02-05 08:54:37 America/New_York
  - added timeout to proc.t

0.63      2013-02-04 08:27:08 America/New_York
  - adjust timout detection and failure in test to be more aggressive

0.62      2012-12-29 09:31:07 America/New_York
  - unstuck tests on MSWin32.  Platform is still unsupported as per the docs.

0.61      2012-11-29 12:19:35 America/New_York
  - Documentation clarification
  - Don't include core prereqs

0.6       2012-10-04 04:48:11 EST5EDT
  - Use Core::close instead of IO::Handle->close
  - Use Core::print instead of IO::Handle->print

0.5       2012-10-02 15:53:35 America/New_York
  - Fix glaring error in the documentation SYNOPSIS

0.4       2012-10-02 15:26:59 America/New_York
  - Fix typeo which broke when passing in hashref to constructor (GH#1)

0.3       2012-10-02 13:42:28 America/New_York
  - documentation fixes
  - re-enable on_error.t for newer perl

0.2       2012-10-02 09:16:57 EST5EDT
  - disabled unreliable test

0.1       2012-10-02 01:20:45 EST5EDT
  - initial version


lib/AnyEvent/Open3/Simple/Process.pm
package AnyEvent::Open3::Simple::Process;

use strict;
use warnings;
use 5.006;

# ABSTRACT: Process run using AnyEvent::Open3::Simple
# VERSION

=head1 DESCRIPTION

This class represents a process being handled by L<AnyEvent::Open3::Simple>.

=cut

sub new
{
  my($class, $pid, $stdin) = @_;
  bless { pid => $pid, stdin => $stdin, user => '' }, $class;
}

=head1 ATTRIBUTES

=head2 pid

 my $pid = $proc->pid;

Return the Process ID of the child process.

=cut

sub pid { shift->{pid} }

=head1 METHODS

=head2 print

 $proc->print(@data);

Write to the subprocess' stdin.

Do NOT use this method if you have passed stdin via the C<$stdin> argument
on the L<AnyEvent::Open3::Simple#run> method.

Currently on (non cygwin) Windows (Strawberry, ActiveState) this method is not
supported, so if you need to send (standard) input to the subprocess, you must pass
it into the L<AnyEvent::Open3::Simple#run> method.

=head2 say

 $proc->say(@data);

Write to the subprocess' stdin, adding a new line at the end.

Do NOT use this method if you have passed stdin via the C<$stdin> argument
on the L<AnyEvent::Open3::Simple#run> method.

Currently on (non cygwin) Windows (Strawberry, ActiveState) this method is not
supported, so if you need to send (standard) input to the subprocess, you must pass
it into the L<AnyEvent::Open3::Simple#run> method.

=cut

if($^O eq 'MSWin32')
{
  require Carp;
  *print = sub { Carp::croak("AnyEvent::Open3::Simple::Process#print is unsupported on this platform") };
  *say = sub { Carp::croak("AnyEvent::Open3::Simple::Process#say is unsupported on this platform") };
}
else
{
  *print = sub {
    my $stdin = shift->{stdin};
    print $stdin @_;
  };
  *say = sub {
    my $stdin = shift->{stdin};
    print $stdin @_, "\n";
  };
}

=head2 close

 $proc->close

Close the subprocess' stdin.

=cut

sub close
{
  CORE::close(shift->{stdin});
}

=head2 user

Version 0.77

 $proc->user($user_data);
 my $user_data = $proc->user;

Get or set user defined data tied to the process object.  Any
Perl data structure may be used.  Useful for persisting data
between callbacks, for example:

 AnyEvent::Open3::Simple->new(
   on_start => sub {
     my($proc) = @_;
     $proc->user({ prefix => '> ' });
   },
   on_stdout => sub {
     my($proc, $line) = @_;
     my $prefix = $proc->user->{prefix};
     say "$prefix$line";
   },
 );

=cut

sub user
{
  my($self, $data) = @_;
  $self->{user} = $data if defined $data;
  $self->{user};
}

1;

=head1 SEE ALSO

=over 4

=item L<AnyEvent::Open3::Simple>

=back

=cut


lib/AnyEvent/Open3/Simple.pm
package AnyEvent::Open3::Simple;

use strict;
use warnings;
use 5.006;
use warnings::register;
use IPC::Open3 qw( open3 );
use Scalar::Util qw( reftype );
use Symbol qw( gensym );
use AnyEvent::Open3::Simple::Process;
use Carp qw( croak );
use File::Temp ();
use constant _is_native_win32 => $^O eq 'MSWin32';
use constant _detect => _is_native_win32() ? 'idle' : 'child';

# ABSTRACT: Interface to open3 under AnyEvent
# VERSION

=head1 SYNOPSIS

 use 5.010;
 use AnyEvent;
 use AnyEvent::Open3::Simple;
 
 my $done = AnyEvent->condvar;
 
 my $ipc = AnyEvent::Open3::Simple->new(
   on_start => sub {
     my $proc = shift;       # isa AnyEvent::Open3::Simple::Process
     my $program = shift;    # string
     my @args = @_;          # list of arguments
     say 'child PID: ', $proc->pid;
   },
   on_stdout => sub {
     my $proc = shift;       # isa AnyEvent::Open3::Simple::Process
     my $line = shift;       # string
     say 'out: ', $string;
   },
   on_stderr => sub {
     my $proc = shift;       # isa AnyEvent::Open3::Simple::Process
     my $line = shift;       # string
     say 'err: ', $line;
   },
   on_exit   => sub {
     my $proc = shift;       # isa AnyEvent::Open3::Simple::Process
     my $exit_value = shift; # integer
     my $signal = shift;     # integer
     say 'exit value: ', $exit_value;
     say 'signal:     ', $signal;
     $done->send;
   },
   on_error => sub {
     my $error = shift;      # the exception thrown by IPC::Open3::open3
     my $program = shift;    # string
     my @args = @_;          # list of arguments
     warn "error: $error";
     $done->send;
   },
 );
 
 $ipc->run('echo', 'hello there');
 $done->recv;

=head1 DESCRIPTION

This module provides an interface to open3 while running under AnyEvent
that delivers data from stdout and stderr as lines are written by the
subprocess.  The interface is reminiscent of L<IPC::Open3::Simple>,
although this module does provides a somewhat different API, so it
cannot be used a drop in replacement for that module.

There are already a number of interfaces for interacting with subprocesses
in the context of L<AnyEvent>, but this one is the most convenient for my
usage.  Note the modules listed in the SEE ALSO section below for other
interfaces that may be more or less appropriate.

=head1 CONSTRUCTOR

Constructor takes a hash or hashref of event callbacks and attributes.
Event callbacks have an C<on_> prefix, attributes do not.

=head2 ATTRIBUTES

=over 4

=item * implementation

The implementation to use for detecting process termination.  This should
be one of C<child>, C<idle> or C<mojo>.  On all platforms except for Microsoft
Windows (but not Cygwin) the default is C<child>.

You can change the default by setting the C<ANYEVENT_OPEN3_SIMPLE>
environment variable, like this:

 % export ANYEVENT_OPEN3_SIMPLE=idle

The C<mojo> implementation is experimental and allows you to use
L<AnyEvent::Open3::Simple> with L<Mojolicious> but without L<EV>
(which is usually required for L<AnyEvent>, L<Mojolicious> interaction).

=back

=head2 EVENTS

These events will be triggered by the subprocess when the run method is
called. Each event callback (except C<on_error>) gets passed in an
instance of L<AnyEvent::Open3::Simple::Process> as its first argument
which can be used to get the PID of the subprocess, or to write to it.
C<on_error> does not get a process object because it indicates an error in
the creation of the process.

Not all of these events will fire depending on the execution of the
child process.  In the very least exactly one of C<on_start> or C<on_error>
will be called.

=over 4

=item * C<on_start> ($proc, $program, @arguments)

Called after the process is created, but before the run method returns
(that is, it does not wait to re-enter the event loop first).

In versions 0.78 and better, this event also gets the program name
and arguments passed into the L<run|AnyEvent::Open3::Simple#run>
method.

=item * C<on_error> ($error, $program, @arguments)

Called when there is an execution error, for example, if you ask
to run a program that does not exist.  No process is passed in
because the process failed to create.  The error passed in is
the error thrown by L<IPC::Open3> (typically a string which begins
with "open3: ...").

In some environments open3 is unable to detect exec errors in the
child, so you may not be able to rely on this event.  It does
seem to work consistently on Perl 5.14 or better though.

Different environments have different ways of handling it when
you ask to run a program that doesn't exist.  On Linux and Cygwin,
this will raise an C<on_error> event, on C<MSWin32> it will
not trigger a C<on_error> and instead cause a normal exit
with a exit value of 1.

In versions 0.77 and better, this event also gets the program name
and arguments passed into the L<run|AnyEvent::Open3::Simple#run>
method.

=item * C<on_stdout> ($proc, $line)

Called on every line printed to stdout by the child process.

=item * C<on_stderr> ($proc, $line)

Called on every line printed to stderr by the child process.

=item * C<on_exit> ($proc, $exit_value, $signal)

Called when the processes completes, either because it called exit,
or if it was killed by a signal.

=item * C<on_success> ($proc)

Called when the process returns zero exit value and is not terminated by a signal.

=item * C<on_signal> ($proc, $signal)

Called when the processes is terminated by a signal.

=item * C<on_fail> ($proc, $exit_value)

Called when the process returns a non-zero exit value.

=back

=cut

sub new
{
  my $default_handler = sub { };
  my $class = shift;
  my $args = (reftype($_[0]) || '') eq 'HASH' ? shift : { @_ };
  my %self;
  croak "stdin passed into AnyEvent::Open3::Simple->new no longer supported" if $args->{stdin};
  croak "raw passed into AnyEvent::Open::Simple->new no longer supported" if $args->{raw};
  $self{$_} = $args->{$_} || $default_handler for qw( on_stdout on_stderr on_start on_exit on_signal on_fail on_error on_success );
  $self{impl} = $args->{implementation}
             || $ENV{ANYEVENT_OPEN3_SIMPLE}
             || _detect();
  croak "unknown implementation $self{impl}" unless $self{impl} =~ /^(idle|child|mojo)$/;
  $self{impl} = _detect()
    if $self{impl} eq 'mojo' && do { require Mojo::Reactor; Mojo::Reactor->detect eq 'Mojo::Reactor::EV' };
  bless \%self, $class;
}

=head1 METHODS

=head2 run

 $ipc->run($program, @arguments);
 $ipc->run($program, @arguments, \$stdin);             # (version 0.76)
 $ipc->run($program, @arguments, \@stdin);             # (version 0.76)
 $ipc->run($program, @arguments, sub {...});           # (version 0.80)
 $ipc->run($program, @arguments, \$stdin, sub {...});  # (version 0.80)
 $ipc->run($program, @arguments, \@stdin, sub {...});  # (version 0.80)

Start the given program with the given arguments.  Returns
immediately.  Any events that have been specified in the
constructor (except for C<on_start>) will not be called until
the process re-enters the event loop.

You may optionally provide the full content of standard input
as a string reference or list reference as the last argument
(or second to last if you are providing a callback below).
If provided as a list reference, it will be joined by new lines
in whatever format is native to your Perl.  Currently on
(non cygwin) Windows (Strawberry, ActiveState) this is the only
way to provide standard input to the subprocess.

Do not mix the use of passing standard input to L<run|AnyEvent::Open3::Simple#run>
and L<AnyEvent::Open3::Simple::Process#print> or L<AnyEvent::Open3::Simple::Process#say>,
otherwise bad things may happen.

In version 0.80 or better, you may provide a callback as the last argument
which is called before C<on_start>, and takes the process object as its only
argument.  For example:

 foreach my $i (1..10)
 {
   $ipc->run($prog, @args, \$stdin, sub {
     my($proc) = @_;
     $proc->user({ iteration => $i });
   });
 }

This is useful for making data accessible to C<$ipc> object's callbacks that may
be out of scope otherwise.

=cut

sub run
{
  croak "run method requires at least one argument"
    unless @_ >= 2;

  my $proc_user = (ref $_[-1] eq 'CODE' ? pop : sub {});

  my $stdin;
  $stdin = pop if ref $_[-1];

  my($self, $program, @arguments) = @_;

  my($child_stdin, $child_stdout, $child_stderr);
  $child_stderr = gensym;

  local *TEMP;
  if(defined $stdin)
  {
    my $file = File::Temp->new;
    $file->autoflush(1);
    $file->print(
      ref($stdin) eq 'ARRAY'
      ? join("\n", @{ $stdin })
      : $$stdin
    );
    $file->seek(0,0);
    open TEMP, '<&=', $file;  ## no critic
    $child_stdin = '<&TEMP';
  }

  if($self->{impl} =~ /^(child|idle)$/)
  {
    require AnyEvent;
    AnyEvent::detect();
    require AnyEvent::Open3::Simple::Idle if $self->{impl} eq 'idle';
  }
  elsif($self->{impl} eq 'mojo')
  {
    require Mojo::Reactor;
    require Mojo::IOLoop;
    require AnyEvent::Open3::Simple::Mojo;
  }

  my $pid = eval { open3 $child_stdin, $child_stdout, $child_stderr, $program, @arguments };

  if(my $error = $@)
  {
    $self->{on_error}->($error, $program, @arguments);
    return;
  }

  my $proc = AnyEvent::Open3::Simple::Process->new($pid, $child_stdin);
  $proc_user->($proc);

  $self->{on_start}->($proc, $program, @arguments);

  my $watcher_stdout;
  my $watcher_stderr;

  my $stdout_callback = sub {
    my $input = <$child_stdout>;
    return unless defined $input;
    $input =~ s/(\015?\012|\015)$//;
    my $ref = $self->{on_stdout};
    ref($ref) eq 'ARRAY' ? push @$ref, $input : $ref->($proc, $input);
  };

  my $stderr_callback = sub {
    my $input = <$child_stderr>;
    return unless defined $input;
    $input =~ s/(\015?\012|\015)$//;
    my $ref = $self->{on_stderr};
    ref($ref) eq 'ARRAY' ? push @$ref, $input : $ref->($proc, $input);
  };

  if(!_is_native_win32() && $self->{impl} =~ /^(idle|child)$/)
  {
    $watcher_stdout = AnyEvent->io(
      fh   => $child_stdout,
      poll => 'r',
      cb   => $stdout_callback,
    ) unless _is_native_win32();

    $watcher_stderr = AnyEvent->io(
      fh   => $child_stderr,
      poll => 'r',
      cb   => $stderr_callback,
    ) unless _is_native_win32();
  }

  my $watcher_child;

  my $end_cb = sub {
    #my($pid, $status) = @_;
    my $status = $_[1];
    my($exit_value, $signal) = ($status >> 8, $status & 127);

    $proc->close;

    # make sure we consume any stdout and stderr which hasn't
    # been consumed yet.  This seems to make on_out.t work on
    # cygwin
    if($self->{raw})
    {
      local $/;
      $self->{on_stdout}->($proc, scalar <$child_stdout>);
      $self->{on_stderr}->($proc, scalar <$child_stderr>);
    }
    else
    {
      while(!eof $child_stdout)
      {
        my $input = <$child_stdout>;
        last unless defined $input;
        $input =~ s/(\015?\012|\015)$//;
        my $ref = $self->{on_stdout};
        ref($ref) eq 'ARRAY' ? push @$ref, $input : $ref->($proc, $input);
      }

      while(!eof $child_stderr)
      {
        my $input = <$child_stderr>;
        last unless defined $input;
        $input =~ s/(\015?\012|\015)$//;
        my $ref = $self->{on_stderr};
        ref($ref) eq 'ARRAY' ? push @$ref, $input : $ref->($proc, $input);
      }
    }

    $self->{on_exit}->($proc, $exit_value, $signal);
    $self->{on_signal}->($proc, $signal) if $signal > 0;
    $self->{on_fail}->($proc, $exit_value) if $exit_value > 0;
    $self->{on_success}->($proc) if $signal == 0 && $exit_value == 0;
    undef $watcher_stdout;
    undef $watcher_stderr;
    undef $watcher_child;
    undef $proc;
  };

  if($self->{impl} eq 'mojo')
  {
    my($selout, $selerr);

    if(_is_native_win32())
    {
      require IO::Select;
      $selout = IO::Select->new($child_stdout);
      $selerr = IO::Select->new($child_stderr);
    }

    my $reactor = Mojo::IOLoop->singleton->reactor;
    my $id;
    $id = Mojo::IOLoop->recurring(0.25 => sub {
      AnyEvent::Open3::Simple::Mojo::_watcher($pid, sub {
        $end_cb->(@_);
        Mojo::IOLoop->remove($id);
        if(_is_native_win32())
        {
          $stdout_callback->() if $selout->can_read(0);
          $stderr_callback->() if $selerr->can_read(0);
        }
        else
        {
          $reactor->remove($child_stdout);
          $reactor->remove($child_stderr);
        }
      });
    });

  }
  elsif($self->{impl} eq 'idle')
  {
    my($selout, $selerr);

    if(_is_native_win32())
    {
      require IO::Select;
      $selout = IO::Select->new($child_stdout);
      $selerr = IO::Select->new($child_stderr);
    }

    $watcher_child = AnyEvent->idle(cb => sub {
      if(_is_native_win32())
      {
        $stdout_callback->() if $selout->can_read(0);
        $stderr_callback->() if $selerr->can_read(0);
      }
      AnyEvent::Open3::Simple::Idle::_watcher($pid, $end_cb);
    });
  }
  else
  {
    $watcher_child = AnyEvent->child(
      pid => $pid,
      cb  => $end_cb,
    );
  }

  $self;
}

1;

=head1 CAVEATS

Some AnyEvent implementations may not work properly with the method
used by AnyEvent::Open3::Simple to wait for the child process to
terminate.  See L<AnyEvent/"CHILD PROCESS WATCHERS"> for details.

This module uses an idle watcher instead of a child watcher to detect
program termination on Microsoft Windows (but not Cygwin).  This is
because the child watchers are unsupported by AnyEvent on Windows.
The idle watcher implementation seems to pass the test suite, but there
may be some traps for the unwary.  There may be other platforms or
event loops where this is the appropriate choice, and you can use the
C<ANYEVENT_OPEN3_SIMPLE> environment variable or the C<implementation>
attribute to force it use an idle watcher instead.  Patches for detecting
environments where idle watchers should be used are welcome and
encouraged.

As of version 0.85, this module works on Windows with L<AnyEvent::Impl::EV>,
L<AnyEvent::Impl::Event> and L<AnyEvent::Impl::Perl> (possibly others),
although in the past they have either not worked or had limitations placed
on them.  Because the author of L<AnyEvent> does not hold the native Windows
port of Perl in high regard: problems such as this may pop up again
in the future and may not be addressed, and may be out of the control of the
author of this module.

Performance for the idle watcher implementation on native Windows (non-Cygwin)
is almost certainly suboptimal, but the author of this module uses it
and finds it useful despite this.

Writing to a subprocesses stdin with L<AnyEvent::Open3::Simple::Process#print>
or L<AnyEvent::Open3::Simple::Process#say> is unsupported on Microsoft
Windows (it does work under Cygwin though).

There are some traps for the unwary relating to buffers and deadlocks,
L<IPC::Open3> is recommended reading.

If you register a call back for C<on_exit>, but not C<on_error> then
use a condition variable to wait for the process to complete as in
this:

 my $cv = AnyEvent->condvar;
 my $ipc = AnyEvent::Open3::Simple->new(
   on_exit => sub { $cv->send },
 );
 $ipc->run('command_not_found');
 $cv->recv;

You might be waiting forever if there is an error starting the
process (if for example you give it a bad command).  To handle
this situation you might use croak on the condition variable
in the event of error:

 my $cv = AnyEvent->condvar;
 my $ipc = AnyEvent::Open3::Simple->new(
   on_exit => sub { $cv->send },
   on_error => sub {
     my $error = shift;
     $cv->croak($error);
   },
 );
 $ipc->run('command_not_found');
 $cv->recv;

This will cause the C<recv> to die, printing a useful diagnostic
if the exception isn't caught somewhere else.

=head1 SEE ALSO

=over 4

=item L<AnyEvent::Open3::Simple::Process>

Represents a process being run by this module, typically passed
into the callbacks.

=item L<AnyEvent::Subprocess>

Alternative to this module.

=item L<AnyEvent::Run>

Alternative to this module.

=back

=cut
